### PR TITLE
Dashboard Scene: Fix issue for Dashboard Datasource and library panel

### PR DESCRIPTION
**Pull request by axelavargas**
**What is this feature?**

We missed the use case of dashboard data source referencing to library panels during the refactor of this PR https://github.com/grafana/grafana/pull/90886 

This was causing a runtime error `Could not find SceneQueryRunner for panel`

This PR adds back the support of this use case :)

https://github.com/user-attachments/assets/dfa55490-e51a-49eb-8de1-cfedd2e39aa6

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx [hunk 1]
```diff
@@ -58,6 +82,25 @@ export class DashboardDatasourceBehaviour extends SceneObjectBase<DashboardDatas
...
+  ) {
+    if (newLibPanel && newLibPanel?.isLoaded && newLibPanel !== oldLibPanel) {
+      const libPanelQueryRunner = getQueryRunnerFor(sourcePanel);
+
+      if (!(libPanelQueryRunner instanceof SceneQueryRunner)) {
+        throw new Error('Could not find SceneQueryRunner for panel');
+      }
+      queryRunner.runQueries();
+    }
+  }
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx:105]: not sure I understand what this code does (what scenario it is for), it works when I am not including it

File: public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx [hunk 2]
```diff
@@ -38,15 +49,28 @@ export class DashboardDatasourceBehaviour extends SceneObjectBase<DashboardDatas
...
-    if (!(panel instanceof VizPanel)) {
+    //check if the source panel is a library panel and wait for it to load
+    if (isLibraryPanel(sourcePanel)) {
+      const libraryPanel = getLibraryPanelBehavior(sourcePanel);
+      if (!libraryPanel) {
+        throw new Error('Could not find LibraryPanelBehavior for panel');
+      }
+      libraryPanelSub = libraryPanel.subscribeToState((newLibPanel, oldLibPanel) => {
+        this.handleLibPanelStateUpdates(newLibPanel, oldLibPanel, queryRunner, sourcePanel);
+      });
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx:69]:
```suggestion
const libPanelBehavior = getLibraryPanelBehavior(sourcePanel);
if (libPanelBehavior) {
    libraryPanelSub = libPanelBehavior.subscribeToState((newLibPanel, oldLibPanel) => {
    this.handleLibPanelStateUpdates(newLibPanel, oldLibPanel, queryRunner, sourcePanel);
});
```

just a shorter form, I am not sure why your passing the queryRunner to handleLibPanelStateUpdates, it's not going to be the correct queryRunner (you can only get this after lib panel has loaded) 

And should you not check for `!libPanelBehavior.state.isLoaded`. ?

**Review by torkelo — COMMENTED** [review]
Not sure the real problem is here (just the exception), if we just do this: 

```diff
 
     if (!sourcePanelQueryRunner) {
-      throw new Error('Could not find SceneQueryRunner for panel');
+      return;
     }
```

It fixes the bug equally well. but I think the problem then moves to DashboardDatasource. It needs some logic to detect lib panels and wait for it to load. 

A related bug is when you have lib panel in a collapsed row and a panel outside referencing it's data. But maybe we can say lib panels does not support that (it looks tricky to solve)

**Comment by mdvictor** [comment]
> Not sure the real problem is here (just the exception), if we just do this:
> 
> ```diff
>  
>      if (!sourcePanelQueryRunner) {
> -      throw new Error('Could not find SceneQueryRunner for panel');
> +      return;
>      }
> ```
> 
> It fixes the bug equally well. but I think the problem then moves to DashboardDatasource. It needs some logic to detect lib panels and wait for it to load.
> 
> A related bug is when you have lib panel in a collapsed row and a panel outside referencing it's data. But maybe we can say lib panels does not support that (it looks tricky to solve)

That's true, the return works, and indeed it might mean moving this to DashboardDS to detect lib panel load. All this code was present in the old implementation and I thought this might fix the issue.

Is that bug you are mentioning limited to only lib panels? I think it's all panels since collapsed rows will not run the panel queries and thus the outside panel will never get updates regardless if it's library or not.

File: public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx (hunk 1 above)
Comment (mdvictor) [reply public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx:105]: This is needed for slow library panels. When a libPanel loads slowly, the panel with dashboardDS will not find any data and show nothing. 

We are not running queries on lib panel, we are only checking for it's existence (assuming that it has loaded the data), and then just run the dashboardDS panel queries to pull data from lib panel

File: public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx [hunk 3]
```diff
@@ -58,6 +82,25 @@ export class DashboardDatasourceBehaviour extends SceneObjectBase<DashboardDatas
...
     };
   }
+
+  private handleLibPanelStateUpdates(
+    newLibPanel: LibraryPanelBehaviorState,
+    oldLibPanel: LibraryPanelBehaviorState,
+    queryRunner: SceneQueryRunner,
+    sourcePanel: VizPanel
+  ) {
+    if (newLibPanel && newLibPanel?.isLoaded && newLibPanel !== oldLibPanel) {
```
Comment (mdvictor) [inline public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx:97]: I think we don't even need the oldPanel here, just checking if `isLoaded` is true should guarantee that we have proper data

File: public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx (hunk 2 above)
Comment (mdvictor) [reply public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx:69]: the passed queryRunner is not the one from the lib panel, but from the dashboardDS panel, so I think it should be alright.

**Comment by mdvictor** [comment]
@torkelo I think we should merge this as is since it's pretty much what we had before. Not entirely sure where we can add this in DashboardDatasource, feels like it should actually stay here.

**Comment by torkelo** [comment]
@mdvictor 
> Is that bug you are mentioning limited to only lib panels? I think it's all panels since collapsed rows will not run the panel queries and thus the outside panel will never get updates regardless if it's library or not.

no it's just for lib panels, for normal panel it works (for the most part), as DashboardDataSource activate the source panel query runner

**Comment by axelavargas** [comment]
@torkelo @mdvictor thanks for the review! I have made small changes to improve the confusion regarding query runner :), and also I tested locally simulating a slow library panel and could confirm the code we added here is indeed needed.

@mdvictor thanks a lot for the guidance with this one ❤️

File: public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx [hunk 4]
```diff
@@ -16,48 +26,79 @@ export class DashboardDatasourceBehaviour extends SceneObjectBase<DashboardDatas
...
+    //check if the source panel is a library panel and wait for it to load
+    if (isLibraryPanel(sourcePanel)) {
+      const libraryPanelBehaviour = getLibraryPanelBehavior(sourcePanel);
+      if (libraryPanelBehaviour) {
+        libraryPanelSub = libraryPanelBehaviour.subscribeToState((newLibPanel) => {
+          this.handleLibPanelStateUpdates(newLibPanel, dashboardDsQueryRunner, sourcePanel);
+        });
+      }
       return;
     }
```
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/dashboard-scene/scene/DashboardDatasourceBehaviour.tsx:70]: also missing a check for if (libraryPanelBehaviour && !libraryPanelBehaviour.state.isLoaded) {

**Review by mdvictor — APPROVED** [review]
Tested again and works like a charm!

**Pull request merged**